File: SampleApp/java/app/src/main/java/jp/fluct/sample/samplefluctsdkapp/banner/BannerRecyclerActivity.java
package jp.fluct.sample.samplefluctsdkapp.banner;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import jp.fluct.fluctsdk.FluctAdView;
import jp.fluct.fluctsdk.FluctErrorCode;
import jp.fluct.sample.samplefluctsdkapp.R;

public class BannerRecyclerActivity extends AppCompatActivity {

    private RecyclerView recycler;
    private FluctAdView adView;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.banner_recycler_activity);

        recycler = findViewById(R.id.recycler);
        recycler.setAdapter(new MyAdapter());
        recycler.setLayoutManager(
            new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        );

        adView =
            new FluctAdView(
                this,
                getString(R.string.banner_test_group_id),
                getString(R.string.banner_test_unit_id),
                getString(R.string.banner_test_ad_size),
                null,
                new FluctAdView.Listener() {
                    @Override
                    public void onLoaded() {
                        toast("onLoaded");
                    }

                    @Override
                    public void onFailedToLoad(@NonNull FluctErrorCode fluctErrorCode) {
                        toast("onFailedToLoad");
                    }

                    @Override
                    public void onLeftApplication() {
                        toast("onLeftApplication");
                    }

                    @Override
                    public void onUnloaded() {
                        toast("onUnloaded");
                    }
                }
            );

        adView.loadAd();
    }

    private void toast(@NonNull String text) {
        Toast.makeText(this, text, Toast.LENGTH_SHORT).show();
    }

    private class MyAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

        @Override
        public int getItemViewType(int position) {
            return position % 50 == 0 ? 1 : 0;
        }

        @NonNull
        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            switch (viewType) {
                case 0:
                    return new MyTextItemVH(parent);
                case 1:
                    return new MyAdItemVH(parent);
                default:
                    throw new IllegalArgumentException("Not implemented!");
            }
        }

        @SuppressLint("SetTextI18n")
        @Override
        public void onBindViewHolder(@NonNull RecyclerView.ViewHolder org, int position) {
            switch (getItemViewType(position)) {
                case 0:
                    {
                        MyTextItemVH vh = (MyTextItemVH) org;
                        vh.text.setText("Position: " + vh.getAdapterPosition());
                        break;
                    }
                case 1:
                    {
                        MyAdItemVH vh = (MyAdItemVH) org;

                        if (vh.container.getChildCount() > 0) {
                            vh.container.removeAllViews();
                        }

                        vh.container.addView(adView);
                        break;
                    }
                default:
                    throw new IllegalArgumentException("Not implemented!");
            }
        }

        @Override
        public int getItemCount() {
            return 1000;
        }
    }

    private class MyTextItemVH extends RecyclerView.ViewHolder {

        public final TextView text;

        public MyTextItemVH(@NonNull ViewGroup parent) {
            this(
                LayoutInflater
                    .from(parent.getContext())
                    .inflate(R.layout.banner_recycler_text_item_layout, parent, false)
            );
        }

        public MyTextItemVH(@NonNull View itemView) {
            super(itemView);
            this.text = itemView.findViewById(R.id.text);
        }
    }

    private class MyAdItemVH extends RecyclerView.ViewHolder {

        public final FrameLayout container;

        public MyAdItemVH(@NonNull ViewGroup parent) {
            this(
                LayoutInflater
                    .from(parent.getContext())
                    .inflate(R.layout.banner_recycler_ad_item_layout, parent, false)
            );
        }

        public MyAdItemVH(@NonNull View itemView) {
            super(itemView);
            this.container = itemView.findViewById(R.id.container);
        }
    }
}

File: SampleApp/kotlin/app/src/main/java/jp/fluct/sample/kotlinapp/banner/BannerRecyclerActivity.kt
package jp.fluct.sample.kotlinapp.banner

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import jp.fluct.fluctsdk.FluctAdView
import jp.fluct.fluctsdk.FluctErrorCode
import jp.fluct.sample.kotlinapp.R

class BannerRecyclerActivity : AppCompatActivity() {

    private val recycler by lazy { findViewById<RecyclerView>(R.id.recycler) }
    private lateinit var adView: FluctAdView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.banner_recycler_activity)

        recycler.adapter = MyAdapter()
        recycler.layoutManager = LinearLayoutManager(
            this,
            LinearLayoutManager.VERTICAL,
            false
        )

        adView = FluctAdView(
            this,
            getString(R.string.banner_test_group_id),
            getString(R.string.banner_test_unit_id),
            getString(R.string.banner_test_ad_size),
            null,
            object : FluctAdView.Listener {

                override fun onUnloaded() {
                    toast("onUnloaded")
                }

                override fun onLeftApplication() {
                    toast("onLeftApplication")
                }

                override fun onLoaded() {
                    toast("onLoaded")
                }

                override fun onFailedToLoad(p0: FluctErrorCode) {
                    toast("onFailedToLoad")
                }

            }
        )

        adView.loadAd()
    }

    private fun toast(text: String) {
        Toast.makeText(this, text, Toast.LENGTH_SHORT).show()
    }

    private inner class MyAdapter : RecyclerView.Adapter<MyVH>() {

        override fun getItemViewType(position: Int): Int {
            return if ((position + 1) % 50 == 0) 0 else 1
        }

        override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): MyVH {
            return when (viewType) {
                0 -> MyAdVH(viewGroup)
                1 -> MyTextVH(viewGroup)
                else -> throw IllegalStateException()
            }
        }

        @SuppressLint("SetTextI18n")
        override fun onBindViewHolder(vh: MyVH, position: Int) {
            when (vh) {
                is MyAdVH -> run {
                    if (vh.container.childCount > 0) {
                        vh.container.removeAllViews()
                    }

                    vh.container.addView(adView)
                }
                is MyTextVH -> vh.text.text = "Position: $position"
            }
        }

        override fun getItemCount(): Int {
            return 1000
        }

    }

}

private sealed class MyVH(itemView: View) : RecyclerView.ViewHolder(itemView)

private class MyAdVH(parent: ViewGroup) : MyVH(
    LayoutInflater.from(parent.context)
        .inflate(R.layout.banner_recycler_ad_item_layout, parent, false)
) {

    val container by lazy { itemView.findViewById<ViewGroup>(R.id.container) }

}

private class MyTextVH(parent: ViewGroup) : MyVH(
    LayoutInflater.from(parent.context)
        .inflate(R.layout.banner_recycler_text_item_layout, parent, false)
) {

    val text by lazy { itemView.findViewById<TextView>(R.id.text) }

}
